### ISSUE TITLE
FileImpl in FS.h can't propagate underlying filesystem errors because of unsigned return values

### ISSUE THREAD

**Issue by ssilverman** [PaulStoffregen/cores#676]
Many of the `FileImpl` functions return `size_t`, however it's very common for the underlying filesystem implementation to return a negative value as an error code. A great example is _LittleFS_. The implementation [here](https://github.com/PaulStoffregen/LittleFS/blob/main/src/LittleFS.h) drops all error codes, for example in the case of LFS_ERR_NOSPC (-28), the very-common "No space left on device" condition.

It's not possible to determine what the problem is just by using the base interfaces from _FS.h_.

It's non-trivial to fix this problem because libraries external to the core depend on the definitions here. _LittleFS_ is a great example.

At least in the `Stream` and `Print` interfaces, there's a way to set and retrieve an associated read or write error code, even through, say, `write()` returns `size_t`.

There are two possible fixes that jump out straight away:
1. Change all unsigned return values to their signed equivalents, for example, `ssize_t` instead of `size_t`.
2. Change `FileImpl` to implement the `Stream` interface; that would provide the necessary error code retrieval mechanism.

In summary, the _FS.h_ interfaces are not completely usable in a production system because it's not possible to detect underlying filesystem errors.
